File: quicksort/Quicksort.ex
defmodule Quicksort do
  import Enum, only: [split_with: 2]

  def sort(list) when length(list) < 2, do: list

  def sort([pivot | tail]) do
    {less, greater} = split_with(tail, &(&1 <= pivot))
    sort(less) ++ [pivot | sort(greater)]
  end
end

IO.inspect(Quicksort.sort([10, 5, 2, 3]))

File: quicksort/Quicksort.rb
def quicksort(array)
    if array.length < 2
      # base case, arrays with 0 or 1 element are already "sorted"
      array
    else
      # recursive case
      pivot = array[0]
      # sub-array of all the elements less than the pivot
      rest = array[1..-1]
      less = rest.select { |i| i <= pivot }
      # sub-array of all the elements greater than the pivot
      greater = rest.select { |i| i > pivot }
  
      quicksort(less) + [pivot] + quicksort(greater)
    end
  end
  
  print quicksort([10, 5, 2, 3])